### PR TITLE
[FLINK-33881]Avoid copy and update value in TtlListState#getUnexpiredOrNull

### PR DESCRIPTION
**Pull request by ljz2051**
## What is the purpose of the change

This pull request optimize TtlListState clean up process by avoiding copy and update value in TtlListState#getUnexpiredOrNull

## Brief change log

  - refactor TtlListState#getUnexpiredOrNull

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.
org.apache.flink.runtime.state.ttl.TtlStateTestBase +  TtlFixedLenElemListStateTestContext/TtlNonFixedLenElemListStateTestContext

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java [hunk 1]
```diff
@@ -109,14 +109,32 @@ public List<TtlValue<T>> getUnexpiredOrNull(@Nonnull List<TtlValue<T>> ttlValues
...
+        }
+
+        List<TtlValue<T>> unexpired = new ArrayList<>(ttlValues.size());
+        for (int i = 0; i < ttlValues.size(); i++) {
+            TtlValue<T> ttlValue = ttlValues.get(i);
+            if (i < firstExpireElementIndex) {
                 // we have to do the defensive copy to update the value
                 unexpired.add(elementSerializer.copy(ttlValue));
+            } else if (i > firstExpireElementIndex) {
+                if (!TtlUtils.expired(ttlValue, ttl, currentTimestamp)) {
```
Comment (Zakelly) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java:134]: How about merge these branches into one?
```
if (i < firstExpireElementIndex || (i > firstExpireElementIndex && !TtlUtils.expired(ttlValue, ttl, currentTimestamp))) {
   //...
}
```

**Review by Zakelly — COMMENTED** [review]
Thanks for this PR @ljz2051 ! Overall LGTM, only one minor comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java (hunk 1 above)
Comment (ljz2051) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java]: Good idea. I've merged it.

**Review by Zakelly — APPROVED** [review]
Thanks for the update, LGTM

**Review by masteryhx — APPROVED** [review]
Thanks for the contribution, LGTM.

**Comment by masteryhx** [comment]
merged.
Just a reminder, remeber to write commit message with component tag next time.

**Pull request closed without merging**